lib/fs.js

```diff
@@ -1013,29 +1013,47 @@ function stat(path, options = { bigint: false }, callback) {
   binding.stat(pathModule.toNamespacedPath(path), options.bigint, req);
 }
 
-function fstatSync(fd, options = { bigint: false }) {
+function fstatSync(fd, options = { bigint: false, throws: true }) {
   validateInt32(fd, 'fd', 0);
   const ctx = { fd };
   const stats = binding.fstat(fd, options.bigint, undefined, ctx);
-  handleErrorFromBinding(ctx);
+  if (options.throws === false) {
+    if (ctx.errno || ctx.error) {
```

Comment:
This would return whatever left inside the stats buffer (probably from the last call) if there's no error which seems to be a bizarre way to leak the stats buffer to the users. Can you just return undefined here `if (options.throws === false)`? (Also the same below)

Reply:
Actually I think it should be

```js
if (ctx.errno || ctx.error) {
  if (options.throws === false) {
    return undefined;
  }
}
handleErrorFromBinding(ctx);
return getStatsFromBinding(stats);
```

Reply:
@joyeecheung I'm not sure I follow.  Which path through the ifs are you concerned about?  If I'm reading your suggested update correctly, it looks like it only differs in calling `handleErrorFromBinding` in the case where no exception is desired and no error occurred (which I expected to be a no-op).

Also, leaking stale stats seems very bad, but I'm not sure I understand how that could occur.  Can you please elaborate?

Reply:
Ah, sorry, I looked into it again and I think you are right, when there's no error and `(options.throws === false)` `handleErrorFromBinding(ctx)` can be skipped.

Reply:
Thanks for checking!  Better safe than sorry. :smile: